tokio/src/sync/mod.rs

```diff
@@ -428,7 +428,7 @@ cfg_sync! {
     pub mod mpsc;
 
     mod mutex;
-    pub use mutex::{Mutex, MutexGuard, TryLockError, OwnedMutexGuard};
+    pub use mutex::{Mutex, MutexGuard, TryLockError, OwnedMutexGuard, MappedMutexGuard};
```

Comment:
I am a little unsure about the profusion of `MutexGuard`s that are now exposed under in `tokio::sync`. We now have a `MutexGuard`, `OwnedMutexGuard`, and `MappedMutexGuard` (and we might also conceivably want a `MappedOwnedMutexGuard`). This seems like a dizzying array of choices to present to users.

Perhaps we should put the "weird" mutex guards in a module or something? @carllerche, any API thoughts?

Reply:
It's definitely a lot. FWIW, parking_lot/lock_api also has a ton of types like this: 

![image](https://user-images.githubusercontent.com/530939/80774874-21cb1280-8b2c-11ea-90b7-52d1ce022d39.png)

Maybe it would be worth having a `sync::guards` module that contains all the various guard types? 

Another option is to have public `sync::mutex`, `sync::rwlock`, etc. modules that each expose their own guard types. This second option seems more in line with other parts of tokio I've seen. See `oneshot::{Sender, Receiver}` and `mpsc::{Sender, Receiver}` for example of that kind of organization strategy.

Reply:
Hmm... @sunjay, do you know if it's _necessary_ for `MappedMutexGuard` to be a distinct type? My thought process is that, as a _user_, I really don't care if a `MutexGuard<U>` was mapped from a `MutexGuard<T>` or came directly from a `Mutex<T>`. Would it be possible to implement something like 
```rust
fn map(MutexGuard<T>, impl Fn(&T) -> &U) -> MutexGuard<U>
```

Reply:
FWIW, I am all for consistency with `parking_lot`/`lock_api` when possible, because having similar API structures to other related crates does help new users learn Tokio's API...but, the image you posted really feels like a bit of a "what _not_ to do"  to me... In some cases, having separate guard types does seem necessary (e.g. `RwLockUpgradeableReadGuard` is definitely a different type from `RwLockReadGuard`, since it has different behaviors), but I'm not convinced this is true for `map`.

Reply:
@hawkw I could replace MutexGuard with an enum that merged the mapped version with the unmapped version. Is the extra runtime cost of matching on the enum okay? Every person on MutexGuard would have a match in it.

Reply:
Hmm, could `MutexGuard` be updated to be:

```rust
pub struct MutexGuard<'a, T> {
    semaphore: &'a Semaphore,
    data: *mut T,
    _p: PhantomData<&mut T>,
}
```

I believe this should be correct and avoid the additional type? What do you think @sunjay.